Add a location to a masterclass

mforner13 opened the pull request:
This PR adds the ability to add a location when creating a masterclass. A masterclass can either take place online or in person. 

If a user chooses online:
- They will add a url and joining instructions for the masterclass
- These will then be added to the draft masterclass, the id for which is in the session

If a user chooses in person:

- They will be prompted to search for a building, address or postcode
- If their search string matches an existing Location in the database, a list of these Location objects will be returned
- If not, the google places API will be queried using their search string and the first three results will be returned
- If the location they select exists in the database, its id will be added to the draft masterclass in session
- If they select a location from google, this will be added as a new location, and the resulting id will be added to the draft masterclass
- A user is then prompted to enter a room number, floor and any instructions for entering the building

TODO:

1. The search in a SQLite databse is automatically case-insensitive, unlike a postgres database. This means that testing the `Location.return_existing_location_or_none()` method against the current test database is not reflective of the app's behaviour. In my next PR I will be updating the tests to use a postgres database to address this.
2. When the 'add location' journey is finished, a user is redirected to the homepage, in a future PR I will add a task list page which a user would be redirected to instead.

mforner13 committed:
Remove unneeded comment

mforner13 committed:
Simplify Masterclass header structure

mforner13 committed:
Choose type of location and redirect

If in person chosen - redirect to location search, if online chosen - redirect to adding online details. Also adds a new base template for form pages which takes a back link and wraps the form in the appropriate main and div elements.

mforner13 committed:
Add online details to a masterclass

If the user chooses an online masterclass, add a url and joining details to the draft masterclass in session and set the is_remote boolean to true. Redirect back to homepage.

mforner13 committed:
Add google maps client with API key

mforner13 committed:
Git ignore flask env

mforner13 committed:
Add flaskenv file to docker-compose

mforner13 committed:
Search for location and display results

If a user enters a string which exists in the location database, return a list of location objects. If not return the first three results from the google places API.

mforner13 committed:
Add location to masterclass

If user has chosen an existing Location object, add the id to the draft masterclass in session. If user chooses a new location from google places, add this to the database and then add its id to the draft masterclass.

mforner13 committed:
Add in person location details to masterclass

Add the room, floor and any instructions for entering the building to an in person masterclass using the draft masterclass id in session.

mforner13 committed:
Add mandatory address field to location fixture

jonodrew commented on app/routes.py:
```diff
@@ -128,4 +128,30 @@ def create_new_content():
+def choose_location_type():
+    if request.method == "POST":
+        if request.form["location-type"] == "online":
+            return redirect(url_for("main_bp.add_online_details"))
```
Another pattern for this could be:
```
location_types = {'online': "main_bp.add_online_details", 'in person': "main_bp.search_for_location"}
if request.method == "POST":
    return redirect(url_for(location_types[request.form["location-type"]]))
```
In this case it doesn't help much, but it's worth considering if you extend this, or if you have another route with more options

mforner13 replied:
Agreed if any more types are added this would be a good plan, but I think it's currently more readable with the two if statements :)

jonodrew commented on app/templates/create-masterclass/location/choose-location-type.html:
```diff
@@ -0,0 +1,43 @@
+
+{% block content %}
+  <div class="govuk-width-container">
+  <a href="/" class="govuk-back-link">Back</a> {# TODO add backlink to tasklist #}
```
TODO probably shouldn't be in production code - consider moving it to an issue tracker (there's a decent enough one on github)

jonodrew commented on app/templates/create-masterclass/location/online-details.html:
```diff
@@ -0,0 +1,41 @@
+  <div class="govuk-width-container">
+  <a href="/create-masterclass/location/type" class="govuk-back-link">Back</a>
+
+  <main class="govuk-main-wrapper " id="main-content" role="main">
```
This is repeated - could you simplify by extending base.html into a "masterclass-location.html" file, and then extending that for these pages?

mforner13 replied:
Good point! Have added a generic `form-page-base.html` template which you pass a back link to. Have added this for all the location pages, and after this PR will change other form template to use it too :)

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -155,3 +155,20 @@ def test_display_my_masterclasses(logged_in_user, test_masterclass_with_details,
+# Add location tests
+
+
+@pytest.mark.parametrize(
```
I'm curious - what happens if the user doesn't select either?

mforner13 replied:
Good point! Have added error validation for location form pages in most recent commit.

jonodrew commented on app/routes.py:
```diff
@@ -145,9 +145,20 @@ def choose_location_type():
 def add_online_details():
+    if request.method == "POST":
+        draft_masterclass = Masterclass.query.get(session["draft_masterclass_id"])
+        draft_masterclass.is_remote = True
```
My personal preference would be to move this to a method on the model and keep the routes/views/controllers as thin as possible, but this is a valid choice

mforner13 replied:
Agreed and have updated. I wasn't sure what to return from the model method as I'm not using the returned value, so returned `self` - what do you think?

jonodrew commented on .sample_flaskenv:
```diff
@@ -0,0 +1,2 @@
+FLASK_APP=masterclasses.py
```
what's this file for?

jonodrew commented on app/templates/create-masterclass/location/search-results.html:
```diff
@@ -0,0 +1,49 @@
+                  </br>
+                  <h1 class="govuk-fieldset__heading"></h1>
+                  <div class="govuk-radios">
+                    {% for result in results[0:3] %}
```
Does this mean you're passing the whole list into the template? Could you slice it in the controller instead?

mforner13 replied:
Yep good spot - am slicing in controller now :)

jonodrew commented on app/templates/create-masterclass/location/in-person-details.html:
```diff
@@ -0,0 +1,50 @@
+                    Instructions for entering the building
+                </label>
+                <span id="more-detail-hint" class="govuk-hint">
+                    What do people know to know in order to enter the building, e.g. can they enter with a cross-government pass?
```
I _think_ the GOV.UK style guide advises against using e.g - can you check?

mforner13 replied:
Yep you're right!

jonodrew reviewed (commented):
A few things to talk through but this is fantastic, clean work. Really well done!

jonodrew commented on app/templates/create-masterclass/location/choose-location-type.html:
```diff
@@ -0,0 +1,33 @@
+{% set back_url="/" %}
```
minor thing, but both here and below I'd recommend using `url_for` rather than hard-coding URLs

mforner13 replied:
Oo yes! Have replaced in all of the location templates :)

jonodrew commented on app/models.py:
```diff
@@ -86,7 +86,11 @@ class Masterclass(db.Model):
+
+    def update_remote_status(self, is_remote: bool):
+        self.is_remote = is_remote
+        return self
```
Returning `self` here returns the whole Masterclass object - is that what you want? I suspect not, so I think it's fine to `return None`. Python methods returns None by default, if you don't have an explicit return statement

jonodrew replied:
Also - I think this is where you might have been taking my advice on thinning down the controller? In which case I'd fully commit and have a method like this: 
```
def set_remote_details(form_data: Dict):
    self.is_remote = True
    self.remote_url = form_data['url']
    self.remote_joining_instructions = form_data['joining_instructions']
    db.session.add(self)
    db.session.commit() 
```
and you'd just pass it `request.form`

Edit - having checked out this branch, it would require a bit of an edit to make it as neat as I'd like. I'll show you what I'm proposing at our next session.

jonodrew commented on app/routes.py:
```diff
@@ -162,7 +164,35 @@ def add_online_details():
 def search_for_location():
+    if request.method == "POST":
+        query = request.form["location"]
+        is_database_data = None
```
I don't think you need this - there's no way `is_database_data` is anything but True or False by the end of this execution, because you've got if/else. You only really need to pre-set the variable if there's a chance it won't be set by the end, for example in an if/elif situation

jonodrew commented on app/routes.py:
```diff
@@ -218,7 +218,16 @@ def location_search_results():
 @login_required
 def add_in_person_location_details():
+    if request.method == "POST":
+        draft_masterclass = Masterclass.query.get(session["draft_masterclass_id"])
```
see previous comment about thinning the controller - we'll discuss approaches at our next session

jonodrew commented on app/routes.py:
```diff
@@ -1,4 +1,14 @@
-from flask import render_template, url_for, flash, redirect, request, Blueprint, session, Response
+from flask import (
```
Since this is just formatting, can we squash it with another change to this file?

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -3,10 +3,8 @@
-from app import db as _db
-from config import *
-from app.models import MasterclassContent, Masterclass, Location, User
+
```
Same as previous comment - any way we can squash this with an existing change?

jonodrew reviewed (changes requested):
This is really good code Miriam. :partying_face:  A few things to discuss and one change.

mforner13 committed:
Add form validation to location form pages

Also Includes formatting of template files.

mforner13 commented:
> This is really good code Miriam. 🥳 A few things to discuss and one change.

Thanks @jonodrew 😄  I've just pushed the suggested changes, as well as a commit which adds error validation to the location form pages. I'm not sure if re-rendering the template and passing any entered form data back in is the most efficient way of doing it though, so would be great to talk about that and thinning the controllers on Wednesday!

mforner13 committed:
Move setting location details logic to models

Set remote and in person location details on a masterclass in the model, rather than controller.

mforner13 commented:
@jonodrew I've just pushed a commit which moves the location setting logic from the controller to the models :)  
I wrote the unit tests, but then realised they were completely replicating what was being tested in the integration tests in `test_masterclasses.py`. E.g. in [test_add_in_person_details](https://github.com/fs-masterclasses/masterclasses/blob/80e02ba4a47eb0aa3cfd3bb6fc86c9203088f5b9/tests/test_masterclasses.py#L174).

Do you think it's okay having just that test and [test_add_online_details](https://github.com/fs-masterclasses/masterclasses/blob/80e02ba4a47eb0aa3cfd3bb6fc86c9203088f5b9/tests/test_masterclasses.py#L233)?

jonodrew commented on app/models.py:
```diff
@@ -102,14 +102,16 @@ def set_location_details(self, data, is_remote):
         self.remote_url = data["url"]
-        self.remote_joining_instructions = data["joining_instructions"]
+        if data["joining_instructions"]:
+            self.remote_joining_instructions = data["joining_instructions"]
```
I can't work out how this works! Where do you set/default to None?

mforner13 replied:
It sets it as none automatically :)

jonodrew reviewed (dismissed):
This looks fantastic. Well done on all this work
